lib/stdlib/src/maps.erl

```diff
@@ -65,6 +67,57 @@ find(_,_) -> erlang:nif_error(undef).
 
 from_list(_) -> erlang:nif_error(undef).
 
+-spec intersect(Map1,Map2) -> Map3 when
+    Map1 :: #{Key => term()},
+    Map2 :: #{term() => Value2},
+    Map3 :: #{Key => Value2}.
+
+intersect(Map1,Map2) when is_map(Map1) andalso
+                          is_map(Map2) ->
+    intersect_with(Map1,Map2,fun intersect_combiner/3);
+intersect(Map1, Map2) ->
+    erlang:error(error_type_two_maps(Map1, Map2),
+                 [Map1,Map2]).
+
+intersect_combiner(_K,_V1,V2) ->
+    V2.
+
+-spec intersect_with(Map1,Map2,Combiner) -> Map3 when
+    Map1 :: #{Key => Value1},
+    Map2 :: #{term() => Value2},
+    Combiner :: fun((Key, Value1, Value2) -> CombineResult),
+    Map3 :: #{Key => CombineResult}.
+
+intersect_with(Map1,Map2,Combiner) when is_map(Map1) andalso
+                                        is_map(Map2) andalso
+                                        is_function(Combiner,3) ->
+    case map_size(Map1) < map_size(Map2) of
+        true ->
+            intersect_with_1(maps:keys(Map1), Map1, Map2, false, Combiner);
```

Comment:
Have you considered using iterators instead of `maps:keys/1`? In the compiler we know the maps are always going to be reasonably small, but the user could pass anything here, so it may be better to do it piece by piece.

Reply:
That is a good point. @jhogberg  Do you have some thoughts about if one of the approaches has a significant execution time performance advantage over the other? Perhaps, a hybrid approach where iterators are used for maps larger than a certain size is preferable if that is the case.

Reply:
I doubt it's worth the extra complexity, but it's worth trying out. Benchmark it :)

Reply:
I have changed the code to use an iterator instead 59c117b. The iterator implementation looks very efficient to me (even for small maps) so I don't think it is worth comparing the performance of the two approaches right now.